core/core.go

```diff
@@ -474,6 +474,23 @@ func (n *IpfsNode) loadBootstrapPeers() ([]peer.PeerInfo, error) {
 func (n *IpfsNode) loadFilesRoot() error {
 	dsk := ds.NewKey("/local/filesroot")
 	pf := func(ctx context.Context, k key.Key) error {
+		ds := n.Repo.Datastore()
+		if old, err := ds.Get(dsk); err == nil && old != nil {
+			_ = n.Pinning.Unpin(ctx, key.Key(old.([]byte)), true)
+		}
+		nnd, err := n.DAG.Get(ctx, k)
+		if err != nil {
+			return err
+		}
+
+		if err := n.Pinning.Pin(ctx, nnd, true); err != nil {
```

Comment:
it looks like no lock is acquired between the unpin + pin. this is unsafe and GC might happen in between.